File: extras/dad_client/dad_client.py
#!/usr/bin/env python3
"""
SPDX-License-Identifier: Apache-2.0
Copyright (c) 2026 Cisco and/or its affiliates.

IPv6 DAD Client Example in Python

This example demonstrates how to use the VPP Python API to:
- Enable/configure IPv6 DAD
- Register for DAD event notifications
- Monitor address state transitions (TENTATIVE -> PREFERRED/DUPLICATE)
- Handle duplicate address detection scenarios
"""

import sys
import signal
import time
from vpp_papi import VPPApiClient

# DAD States (from ip6_dad.h)
DAD_STATE_IDLE = 0
DAD_STATE_TENTATIVE = 1
DAD_STATE_PREFERRED = 2
DAD_STATE_DUPLICATE = 3

STATE_NAMES = {
    DAD_STATE_IDLE: "IDLE",
    DAD_STATE_TENTATIVE: "TENTATIVE",
    DAD_STATE_PREFERRED: "PREFERRED",
    DAD_STATE_DUPLICATE: "DUPLICATE"
}

class DADClient:
    """Simple DAD event monitoring client"""

    def __init__(self):
        self.vpp = None
        self.running = True

    def connect(self):
        """Connect to VPP"""
        print("Connecting to VPP...")
        self.vpp = VPPApiClient(apifiles=[/usr/share/vpp/api])
        self.vpp.connect("dad_client")
        print("✓ Connected to VPP")

    def disconnect(self):
        """Disconnect from VPP"""
        if self.vpp:
            print("\nDisconnecting from VPP...")
            self.vpp.disconnect()
            print("✓ Disconnected")

    def enable_dad(self, transmits=1, delay=1.0):
        """
        Enable DAD with configuration

        Args:
            transmits: Number of NS transmissions (1-10)
            delay: Delay between transmissions in seconds (0.1-10.0)
        """
        print(f"\nEnabling DAD (transmits={transmits}, delay={delay}s)...")
        try:
            reply = self.vpp.api.ip6_dad_enable_disable(
                enable=True,
                dad_transmits=transmits,
                dad_retransmit_delay=delay
            )
            if reply.retval == 0:
                print("✓ DAD enabled")
            else:
                print(f"✗ DAD enable failed: retval={reply.retval}")
                return False
        except Exception as e:
            print(f"✗ DAD enable error: {e}")
            return False
        return True

    def register_for_events(self):
        """Register to receive DAD event notifications"""
        print("\nRegistering for DAD events...")
        try:
            reply = self.vpp.api.want_ip6_dad_events(enable=True)
            if reply.retval == 0:
                print("✓ Registered for DAD events")
            else:
                print(f"✗ Registration failed: retval={reply.retval}")
                return False
        except Exception as e:
            print(f"✗ Registration error: {e}")
            return False
        return True

    def handle_dad_event(self, event):
        """
        Handle DAD event notification

        Args:
            event: ip6_dad_event message from VPP
        """
        state = event.state
        state_name = STATE_NAMES.get(state, "UNKNOWN")
        sw_if_index = event.sw_if_index
        address = event.address
        dad_count = event.dad_count
        dad_transmits = event.dad_transmits

        print(f"\n📢 DAD Event Received:")
        print(f"   Interface : sw_if_index {sw_if_index}")
        print(f"   Address   : {address}")
        print(f"   State     : {state_name} ({state})")
        print(f"   Progress  : {dad_count}/{dad_transmits}")

        # Handle different states
        if state == DAD_STATE_TENTATIVE:
            if dad_count == 0:
                print(f"   → Initial TENTATIVE state (before first NS)")
            else:
                print(f"   → TENTATIVE after NS #{dad_count}")

        elif state == DAD_STATE_PREFERRED:
            print(f"   → SUCCESS: Address is now usable (PREFERRED)")
            print(f"   ✓ No duplicate detected after {dad_count} probes")

        elif state == DAD_STATE_DUPLICATE:
            print(f"   → FAILURE: Duplicate address detected!")
            print(f"   ⚠ IMPORTANT: IP address REMAINS CONFIGURED but in DUPLICATE state")
            print(f"   ⚠ Application should decide what to do:")
            print(f"      1. Log error / send alert")
            print(f"      2. Generate new IPv6 address (e.g., privacy extensions)")
            print(f"      3. Remove address manually")
            print(f"      4. Disable interface")
            print(f"      5. Failover to backup address")

            # Example: Application-level decision
            print(f"\n   📋 Suggested action: Generate new address with privacy extensions")

    def monitor_events(self, duration=30):
        """
        Monitor DAD events for specified duration

        Args:
            duration: Duration in seconds to monitor (0 = infinite)
        """
        print(f"\n👀 Monitoring DAD events" + (f" for {duration}s..." if duration > 0 else "..."))
        print("   (Press Ctrl+C to stop)")

        start_time = time.time()

        try:
            while self.running:
                # Check for events
                try:
                    events = self.vpp.message_queue.get(timeout=1.0)

                    # Handle events
                    if hasattr(events, __iter__):
                        for event in events:
                            if hasattr(event, _fields) and state in event._fields:
                                self.handle_dad_event(event)
                    elif hasattr(events, _fields) and state in events._fields:
                        self.handle_dad_event(events)

                except:
                    # Timeout or no events
                    pass

                # Check duration
                if duration > 0 and (time.time() - start_time) >= duration:
                    break

        except KeyboardInterrupt:
            print("\n\n⚠ Interrupted by user")

        print("\n✓ Monitoring stopped")

    def run(self):
        """Main execution"""
        print("="*70)
        print("  IPv6 DAD Client Example (Python)")
        print("="*70)

        try:
            # Connect to VPP
            self.connect()

            # Enable DAD with 2 transmits, 1.0s delay
            if not self.enable_dad(transmits=2, delay=1.0):
                return 1

            # Register for events
            if not self.register_for_events():
                return 1

            # Monitor events for 30 seconds
            # In real application, this would run continuously
            self.monitor_events(duration=30)

        except Exception as e:
            print(f"\n✗ Error: {e}")
            return 1
        finally:
            self.disconnect()

        print("\n" + "="*70)
        print("  Example completed")
        print("="*70)
        return 0


def signal_handler(sig, frame):
    """Handle Ctrl+C gracefully"""
    print("\n\nReceived signal, exiting...")
    sys.exit(0)


if __name__ == "__main__":
    # Setup signal handler
    signal.signal(signal.SIGINT, signal_handler)

    # Run client
    client = DADClient()
    sys.exit(client.run())

File: extras/dad_client/dad_client.c
/*
 * SPDX-License-Identifier: Apache-2.0
 * Copyright (c) 2026 Cisco and/or its affiliates.
 */

/**
 * @file dad_client.c
 * @brief IPv6 DAD Client Example using VAPI
 *
 * This is a structural example demonstrating how to integrate with
 * VPP's IPv6 DAD event notification system using VAPI.
 *
 * Note: This example shows the integration pattern. Actual implementation
 * may require adjustments based on your VPP version and build environment.
 *
 * Key concepts demonstrated:
 * - VAPI connection setup
 * - DAD configuration via API
 * - Event registration
 * - Event callback handling
 * - Proper cleanup
 *
 * Expected event sequence for a successful DAD:
 * 1. TENTATIVE (dad_count=0) - Initial state
 * 2. TENTATIVE (dad_count=1..N) - After each NS probe
 * 3. PREFERRED - Address validated, ready to use
 *
 * Or in case of conflict:
 * 1. TENTATIVE (dad_count=0) - Initial state
 * 2. DUPLICATE - Conflict detected, IP remains configured!
 *
 * Important: When DUPLICATE is detected, the IP address remains configured.
 * The application must decide how to handle this.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <signal.h>

/* Note: In a real implementation, these would be:
 * #include <vapi/vapi.h>
 * #include <vapi/ip6_dad.api.vapi.h>
 */

/* DAD state constants (from VPP API) */
#define DAD_STATE_IDLE	    0
#define DAD_STATE_TENTATIVE 1
#define DAD_STATE_PREFERRED 2
#define DAD_STATE_DUPLICATE 3

/* Global flag for clean shutdown */
static volatile int keep_running = 1;

/**
 * Signal handler for clean shutdown
 */
static void
signal_handler (int signum)
{
  (void) signum;
  keep_running = 0;
}

/**
 * Convert DAD state to string
 */
static const char *
dad_state_to_string (int state)
{
  switch (state)
    {
    case DAD_STATE_IDLE:
      return "IDLE";
    case DAD_STATE_TENTATIVE:
      return "TENTATIVE";
    case DAD_STATE_PREFERRED:
      return "PREFERRED";
    case DAD_STATE_DUPLICATE:
      return "DUPLICATE";
    default:
      return "UNKNOWN";
    }
}

/**
 * Handle DAD event notification
 *
 * This callback is invoked by VAPI when a DAD event is received.
 *
 * @param event Pointer to the DAD event structure
 */
static void
handle_dad_event (/* vapi_msg_ip6_dad_event *event */ void *event_data)
{
  /* In a real implementation, event_data would be cast to:
   * vapi_msg_ip6_dad_event *event = (vapi_msg_ip6_dad_event *)event_data;
   *
   * For this structural example, we'll simulate the event structure:
   */
  struct
  {
    int sw_if_index;
    char address[40]; /* IPv6 address as string */
    int state;
    int dad_count;
  } *event = event_data;

  /* Display event details */
  printf ("\n[DAD Event]\n");
  printf ("  Interface:  sw_if_index=%d\n", event->sw_if_index);
  printf ("  Address:    %s\n", event->address);
  printf ("  State:      %s (%d)\n", dad_state_to_string (event->state), event->state);
  printf ("  DAD Count:  %d\n", event->dad_count);

  /* Interpret the event */
  switch (event->state)
    {
    case DAD_STATE_TENTATIVE:
      if (event->dad_count == 0)
	{
	  printf ("  → DAD process started (initial TENTATIVE)\n");
	}
      else
	{
	  printf ("  → NS probe #%d sent, no conflict detected yet\n", event->dad_count);
	}
      break;

    case DAD_STATE_PREFERRED:
      printf ("  → ✓ DAD succeeded! Address is PREFERRED and ready to "
	      "use\n");
      break;

    case DAD_STATE_DUPLICATE:
      printf ("  → ✗ DUPLICATE DETECTED!\n");
      printf ("  → ⚠️  Address remains configured but should not be used\n");
      printf ("  → Application must decide: remove address or take other action\n");

      /* Application logic for handling duplicates goes here.
       * Options include:
       * 1. Remove the address via API
       * 2. Generate a new address (e.g., privacy extensions)
       * 3. Alert the user/administrator
       * 4. Log the event for security monitoring
       */
      break;

    default:
      printf ("  → Unknown state\n");
      break;
    }
}

/**
 * Main program
 */
int
main (int argc, char *argv[])
{
  (void) argc;
  (void) argv;

  /* In a real implementation, you would:
   * 1. Initialize VAPI
   * 2. Connect to VPP
   * 3. Enable DAD via API
   * 4. Register for events
   * 5. Enter event loop
   * 6. Cleanup on exit
   */

  printf ("============================================================\n");
  printf ("VPP IPv6 DAD Event Monitor (Structural Example)\n");
  printf ("============================================================\n");
  printf ("\n");
  printf ("This is a structural example showing VAPI integration "
	  "patterns.\n");
  printf ("For a working example, see the Python implementation: "
	  "dad_client.py\n");
  printf ("\n");

  /* Setup signal handlers */
  signal (SIGINT, signal_handler);
  signal (SIGTERM, signal_handler);

  printf ("Step-by-step integration guide:\n");
  printf ("\n");

  printf ("1. Initialize VAPI connection\n");
  printf ("   Example:\n");
  printf ("     vapi_ctx_t ctx;\n");
  printf ("     vapi_error_e rv = vapi_ctx_alloc(&ctx);\n");
  printf ("     if (rv != VAPI_OK) { /* handle error */ }\n");
  printf ("\n");

  printf ("2. Connect to VPP\n");
  printf ("   Example:\n");
  printf ("     rv = vapi_connect(ctx, \"dad_client\", NULL,\n");
  printf ("                       DEFAULT_MAX_OUTSTANDING_REQUESTS,\n");
  printf ("                       DEFAULT_RESPONSE_QUEUE_SIZE);\n");
  printf ("     if (rv != VAPI_OK) { /* handle error */ }\n");
  printf ("\n");

  printf ("3. Enable and configure DAD\n");
  printf ("   Example:\n");
  printf ("     vapi_msg_ip6_dad_enable_disable *msg =\n");
  printf ("       vapi_alloc_ip6_dad_enable_disable(ctx);\n");
  printf ("     msg->payload.enable = 1;\n");
  printf ("     msg->payload.dad_transmits = 1;        /* RFC 4862 "
	  "default */\n");
  printf ("     msg->payload.dad_retransmit_delay = 1.0; /* 1 second */\n");
  printf ("     vapi_ip6_dad_enable_disable(ctx, msg, callback, NULL);\n");
  printf ("\n");

  printf ("4. Register for event notifications\n");
  printf ("   Example:\n");
  printf ("     vapi_msg_want_ip6_dad_events *evt_msg =\n");
  printf ("       vapi_alloc_want_ip6_dad_events(ctx);\n");
  printf ("     evt_msg->payload.enable = 1;\n");
  printf ("     evt_msg->payload.pid = getpid();\n");
  printf ("     vapi_want_ip6_dad_events(ctx, evt_msg, callback, NULL);\n");
  printf ("\n");

  printf ("5. Setup event handler\n");
  printf ("   Example:\n");
  printf ("     vapi_set_event_cb(ctx, vapi_msg_id_ip6_dad_event,\n");
  printf ("                       handle_dad_event, NULL);\n");
  printf ("\n");

  printf ("6. Enter event loop\n");
  printf ("   Example:\n");
  printf ("     while (keep_running) {\n");
  printf ("       rv = vapi_dispatch(ctx);\n");
  printf ("       if (rv != VAPI_OK) { /* handle error */ }\n");
  printf ("     }\n");
  printf ("\n");

  printf ("7. Cleanup\n");
  printf ("   Example:\n");
  printf ("     /* Unregister from events */\n");
  printf ("     evt_msg->payload.enable = 0;\n");
  printf ("     vapi_want_ip6_dad_events(ctx, evt_msg, NULL, NULL);\n");
  printf ("     vapi_disconnect(ctx);\n");
  printf ("     vapi_ctx_free(ctx);\n");
  printf ("\n");

  printf ("============================================================\n");
  printf ("Event Handling Example\n");
  printf ("============================================================\n");
  printf ("\n");
  printf ("Simulating DAD events for address 2001:db8::1:\n");
  printf ("\n");

  /* Simulate event sequence */
  struct
  {
    int sw_if_index;
    char address[40];
    int state;
    int dad_count;
  } simulated_events[] = {
    { 1, "2001:db8::1", DAD_STATE_TENTATIVE, 0 },
    { 1, "2001:db8::1", DAD_STATE_TENTATIVE, 1 },
    { 1, "2001:db8::1", DAD_STATE_PREFERRED, 1 },
  };

  for (int i = 0; i < 3; i++)
    {
      handle_dad_event (&simulated_events[i]);
      sleep (1);
    }

  printf ("\n");
  printf ("============================================================\n");
  printf ("Duplicate Detection Example\n");
  printf ("============================================================\n");
  printf ("\n");
  printf ("Simulating duplicate detection for address 2001:db8::2:\n");
  printf ("\n");

  struct
  {
    int sw_if_index;
    char address[40];
    int state;
    int dad_count;
  } conflict_events[] = {
    { 1, "2001:db8::2", DAD_STATE_TENTATIVE, 0 },
    { 1, "2001:db8::2", DAD_STATE_DUPLICATE, 0 },
  };

  for (int i = 0; i < 2; i++)
    {
      handle_dad_event (&conflict_events[i]);
      sleep (1);
    }

  printf ("\n");
  printf ("============================================================\n");
  printf ("Summary\n");
  printf ("============================================================\n");
  printf ("\n");
  printf ("This structural example demonstrates:\n");
  printf ("  ✓ Event callback structure\n");
  printf ("  ✓ State interpretation\n");
  printf ("  ✓ Duplicate handling considerations\n");
  printf ("\n");
  printf ("Key Points:\n");
  printf ("  • IP addresses remain configured in DUPLICATE state\n");
  printf ("  • Applications must implement their own duplicate handling "
	  "policy\n");
  printf ("  • Events provide complete state transition visibility\n");
  printf ("\n");
  printf ("For a complete, working implementation, see: dad_client.py\n");
  printf ("\n");

  return 0;
}
